File: Module/PlayWays/src/main/java-room-grab/com/module/playways/grab/room/songmanager/GrabSongManageFragment.java
package com.module.playways.grab.room.songmanager;

import android.os.Bundle;
import android.os.Handler;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.view.Gravity;
import android.view.View;
import android.view.animation.Animation;
import android.view.animation.TranslateAnimation;
import android.widget.ImageView;
import android.widget.PopupWindow;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.common.base.BaseActivity;
import com.common.base.BaseFragment;
import com.common.base.FragmentDataListener;
import com.common.image.fresco.FrescoWorker;
import com.common.image.model.ImageFactory;
import com.common.image.model.oss.OssImgFactory;
import com.common.log.MyLog;
import com.common.utils.FragmentUtils;
import com.common.utils.ImageUtils;
import com.common.utils.U;
import com.common.view.DebounceViewClickListener;
import com.common.view.ex.ExLinearLayout;
import com.common.view.ex.ExTextView;
import com.facebook.drawee.drawable.ScalingUtils;
import com.facebook.drawee.view.SimpleDraweeView;
import com.module.playways.grab.room.GrabRoomData;
import com.module.playways.grab.room.inter.IGrabSongManageView;
import com.module.playways.grab.room.songmanager.tags.GrabSongTagsView;
import com.component.busilib.friends.SpecialModel;
import com.module.playways.grab.room.songmanager.tags.GrabTagsAdapter;
import com.module.playways.grab.room.songmanager.view.GrabEditView;
import com.module.playways.room.song.fragment.GrabSearchSongFragment;
import com.module.playways.room.song.model.SongModel;
import com.module.playways.R;
import com.orhanobut.dialogplus.DialogPlus;
import com.orhanobut.dialogplus.ViewHolder;
import com.scwang.smartrefresh.layout.SmartRefreshLayout;
import com.scwang.smartrefresh.layout.api.RefreshLayout;
import com.scwang.smartrefresh.layout.listener.OnRefreshLoadMoreListener;

import java.util.List;

public class GrabSongManageFragment extends BaseFragment implements IGrabSongManageView {
    public final static String TAG = "GrabSongManageFragment";

    GrabRoomData mRoomData;

    ExTextView mSearchSongIv;

    ExTextView mTvSelectedSong;

    ExTextView mEditRoomName;

    ExLinearLayout mFlSongListContainer;

    SmartRefreshLayout mRefreshLayout;

    RecyclerView mRecyclerView;

    SimpleDraweeView mTvSelectedTag;

    TextView mTvFinish;

    ImageView mIvArrow;

    ManageSongAdapter mManageSongAdapter;

    GrabSongManagePresenter mGrabSongManagePresenter;

    GrabSongTagsView mGrabSongTagsView;

    PopupWindow mPopupWindow;

    RelativeLayout mRlContent;

    DialogPlus mEditRoomDialog;

    Handler mUiHandler = new Handler();

    int mSpecialModelId;

    @Override
    public int initView() {
        return R.layout.grab_song_manage_fragment_layout;
    }

    @Override
    public void initData(@Nullable Bundle savedInstanceState) {
        mGrabSongManagePresenter = new GrabSongManagePresenter(this, mRoomData);
        addPresent(mGrabSongManagePresenter);

        mIvArrow = (ImageView) mRootView.findViewById(R.id.iv_arrow);
        mSearchSongIv = mRootView.findViewById(R.id.search_song_iv);
        mTvSelectedSong = (ExTextView) mRootView.findViewById(R.id.tv_selected_song);
        mEditRoomName = (ExTextView) mRootView.findViewById(R.id.edit_room_name);
        mFlSongListContainer = (ExLinearLayout) mRootView.findViewById(R.id.fl_song_list_container);
        mRefreshLayout = (SmartRefreshLayout) mRootView.findViewById(R.id.refreshLayout);
        mRecyclerView = (RecyclerView) mRootView.findViewById(R.id.recycler_view);
        mTvSelectedTag = mRootView.findViewById(R.id.selected_tag);
        mTvFinish = (TextView) mRootView.findViewById(R.id.tv_finish);
        mRlContent = (RelativeLayout) mRootView.findViewById(R.id.rl_content);

        mRecyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        mManageSongAdapter = new ManageSongAdapter();
        mRecyclerView.setAdapter(mManageSongAdapter);

        mRefreshLayout.setEnableRefresh(false);
        mRefreshLayout.setEnableLoadMore(true);
        mRefreshLayout.setEnableLoadMoreWhenContentNotFull(true);
        mRefreshLayout.setEnableOverScrollDrag(false);
        mRefreshLayout.setOnRefreshLoadMoreListener(new OnRefreshLoadMoreListener() {
            @Override
            public void onLoadMore(@NonNull RefreshLayout refreshLayout) {
                mGrabSongManagePresenter.getPlayBookList();
            }

            @Override
            public void onRefresh(@NonNull RefreshLayout refreshLayout) {

            }
        });

        DefaultItemAnimator defaultItemAnimator = new DefaultItemAnimator();
        defaultItemAnimator.setAddDuration(50);
        defaultItemAnimator.setRemoveDuration(50);
        mRecyclerView.setItemAnimator(defaultItemAnimator);

        initListener();
        mGrabSongManagePresenter.getPlayBookList();

        if (mRoomData.getSpecialModel() != null) {
            setTagTv(mRoomData.getSpecialModel());
        }

        mUiHandler.postDelayed(() -> {
            TranslateAnimation animation = new TranslateAnimation(Animation.RELATIVE_TO_SELF, 1.0f, Animation.RELATIVE_TO_SELF, 0.0f,
                    Animation.RELATIVE_TO_SELF, 0, Animation.RELATIVE_TO_SELF, 0);
            animation.setDuration(400);
            animation.setRepeatMode(Animation.REVERSE);
            animation.setFillAfter(true);
            mRlContent.startAnimation(animation);
            mRlContent.setVisibility(View.VISIBLE);
        }, 50);
    }

    private void setTagTv(SpecialModel specialModel) {
        mRoomData.setSpecialModel(specialModel);
        mRoomData.setTagId(specialModel.getTagID());

        mSpecialModelId = specialModel.getTagID();

        FrescoWorker.loadImage(mTvSelectedTag, ImageFactory.newPathImage(specialModel.getBgImage3())
                .setScaleType(ScalingUtils.ScaleType.FIT_XY)
                .addOssProcessors(OssImgFactory.newResizeBuilder().setW(ImageUtils.SIZE.SIZE_640.getW()).build())
                .build());
    }

    @Override
    public void changeTagSuccess(SpecialModel specialModel) {
        setTagTv(specialModel);
        if (mPopupWindow != null) {
            mPopupWindow.dismiss();
        }
    }

    @Override
    public void showNum(int num) {
        mTvSelectedSong.setText(String.format("已点%d首", num));
    }

    @Override
    public void deleteSong(GrabRoomSongModel grabRoomSongModel) {
        mManageSongAdapter.deleteSong(grabRoomSongModel);
    }

    @Override
    public void hasMoreSongList(boolean hasMore) {
        mRefreshLayout.setEnableLoadMore(hasMore);
        mRefreshLayout.finishLoadMore();
    }

    @Override
    public void showTagList(List<SpecialModel> specialModelList) {
        int height = U.getDisplayUtils().dip2px(specialModelList.size() > 4 ? 190 : (55 * (specialModelList.size() - 1)));
        if (mGrabSongTagsView != null) {
            mGrabSongTagsView.setSpecialModelList(specialModelList);
            mPopupWindow.setHeight(height);
        }

        int[] location = new int[2];
        mTvSelectedTag.getLocationOnScreen(location);
//        mPopupWindow.showAtLocation(mTvSelectedTag, Gravity.NO_GRAVITY, location[0], location[1] - height - U.getDisplayUtils().dip2px(5));
        mPopupWindow.showAsDropDown(mTvSelectedTag);
        mIvArrow.setBackground(U.getDrawable(R.drawable.fz_shuxing_shang));
    }

    @Override
    public void updateSongList(List<GrabRoomSongModel> grabRoomSongModelsList) {
        mManageSongAdapter.setDataList(grabRoomSongModelsList);
    }

    private void initListener() {
        mEditRoomName.setOnClickListener(new DebounceViewClickListener() {
            @Override
            public void clickValid(View v) {
                showEditRoomDialog();
            }
        });


        mSearchSongIv.setOnClickListener(new DebounceViewClickListener() {
            @Override
            public void clickValid(View v) {
                U.getFragmentUtils().addFragment(FragmentUtils.newAddParamsBuilder((BaseActivity) getContext(), GrabSearchSongFragment.class)
                        .setAddToBackStack(true)
                        .setHasAnimation(true)
                        .setFragmentDataListener(new FragmentDataListener() {
                            @Override
                            public void onFragmentResult(int requestCode, int resultCode, Bundle bundle, Object obj) {
                                if (requestCode == 0 && resultCode == 0 && obj != null) {
                                    SongModel model = (SongModel) obj;
                                    MyLog.d(TAG, "onFragmentResult" + " model=" + model);
                                    mGrabSongManagePresenter.addSong(model);
                                }
                            }
                        })
                        .build());
            }
        });

        mTvSelectedTag.setOnClickListener(v -> {
            if (mGrabSongTagsView == null) {
                mGrabSongTagsView = new GrabSongTagsView(getContext());

                mGrabSongTagsView.setOnTagClickListener(new GrabTagsAdapter.OnTagClickListener() {
                    @Override
                    public void onClick(SpecialModel specialModel) {
                        mGrabSongManagePresenter.changeMusicTag(specialModel, mRoomData.getGameId());
                    }

                    @Override
                    public void dismissDialog() {
                        if (mPopupWindow != null) {
                            mPopupWindow.dismiss();
                        }
                    }
                });
                mPopupWindow = new PopupWindow(mGrabSongTagsView);
                mPopupWindow.setWidth(mTvSelectedTag.getWidth());
                mPopupWindow.setOutsideTouchable(true);
                mPopupWindow.setFocusable(true);

                mPopupWindow.setOnDismissListener(new PopupWindow.OnDismissListener() {
                    @Override
                    public void onDismiss() {
                        mIvArrow.setBackground(U.getDrawable(R.drawable.fz_shuxing_xia));
                    }
                });
            }

            mGrabSongTagsView.setCurSpecialModel(mSpecialModelId);

            if (mPopupWindow.isShowing()) {
                mPopupWindow.dismiss();
            } else {
                mGrabSongManagePresenter.getTagList();
            }
        });

        mTvFinish.setOnClickListener(v -> {
            finish();
        });

        mRootView.setOnClickListener(v -> {
            finish();
        });

        mRlContent.setOnClickListener(v -> {

        });

        mManageSongAdapter.setOnClickDeleteListener(grabRoomSongModel -> {
            mGrabSongManagePresenter.deleteSong(grabRoomSongModel);
        });

        mManageSongAdapter.setGrabRoomData(mRoomData);
    }

    private void showEditRoomDialog() {
        GrabEditView grabEditView = new GrabEditView(getContext());
        grabEditView.setListener(new GrabEditView.Listener() {
            @Override
            public void onClickCancel() {
                if (mEditRoomDialog != null) {
                    mEditRoomDialog.dismiss();
                }
            }

            @Override
            public void onClickSave(String roomName) {
                if (!TextUtils.isEmpty(roomName)) {
                    // TODO: 2019/4/18 修改房间名
                    mEditRoomDialog.dismiss(false);
                    mGrabSongManagePresenter.updateRoomName(mRoomData.getGameId(), roomName);
                } else {
                    // TODO: 2019/4/18 房间名为空
                    U.getToastUtil().showShort("输入的房间名为空");
                }
            }
        });

        mEditRoomDialog = DialogPlus.newDialog(getContext())
                .setContentHolder(new ViewHolder(grabEditView))
                .setContentBackgroundResource(R.color.transparent)
                .setOverlayBackgroundResource(R.color.black_trans_50)
                .setExpanded(false)
                .setGravity(Gravity.CENTER)
                .create();
        mEditRoomDialog.show();
    }

    @Override
    public void destroy() {
        super.destroy();
        mRlContent.clearAnimation();
        if (mEditRoomDialog != null) {
            mEditRoomDialog.dismiss(false);
        }
    }

    @Override
    public void setData(int type, @Nullable Object data) {
        super.setData(type, data);
        if (type == 0) {
            mRoomData = (GrabRoomData) data;
        }
    }

    @Override
    protected boolean onBackPressed() {
        return super.onBackPressed();
    }

    @Override
    public boolean useEventBus() {
        return false;
    }
}

File: Module/PlayWays/src/main/java-songmanage/com/module/playways/songmanager/fragment/GrabSongManageFragment.kt
package com.module.playways.songmanager.fragment

import android.graphics.Rect
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.support.v4.view.PagerAdapter
import android.support.v4.view.ViewPager
import android.text.TextUtils
import android.view.Gravity
import android.view.View
import android.view.ViewGroup

import com.common.base.BaseActivity
import com.common.base.BaseFragment
import com.common.base.FragmentDataListener
import com.common.log.MyLog
import com.common.utils.FragmentUtils
import com.common.utils.U
import com.common.view.DebounceViewClickListener
import com.common.view.ex.ExTextView
import com.common.view.titlebar.CommonTitleBar
import com.common.view.viewpager.SlidingTabLayout
import com.module.playways.R
import com.module.playways.grab.room.GrabRoomData
import com.module.playways.songmanager.SongManagerActivity
import com.module.playways.songmanager.event.AddSongEvent
import com.module.playways.songmanager.event.SongNumChangeEvent
import com.module.playways.songmanager.model.RecommendTagModel
import com.module.playways.songmanager.presenter.GrabSongManagePresenter
import com.module.playways.songmanager.view.GrabEditRoomNameView
import com.module.playways.songmanager.view.GrabExistSongManageView
import com.module.playways.songmanager.view.GrabSongWishView
import com.module.playways.songmanager.view.ISongManageView
import com.module.playways.songmanager.view.RecommendSongView
import com.module.playways.room.song.fragment.GrabSearchSongFragment
import com.module.playways.room.song.model.SongModel
import com.orhanobut.dialogplus.DialogPlus
import com.orhanobut.dialogplus.OnDismissListener
import com.orhanobut.dialogplus.ViewHolder

import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode

/**
 * 一唱到底，房主点歌或房成员点歌
 */
class GrabSongManageFragment : BaseFragment(), ISongManageView {
    lateinit var mCommonTitleBar: CommonTitleBar
    lateinit var mSearchSongIv: ExTextView
    lateinit var mTagTab: SlidingTabLayout
    lateinit var mViewpager: ViewPager

    lateinit var mPagerAdapter: PagerAdapter

    private var mGrabSongManageView: GrabExistSongManageView? = null
    private var mGrabSongWishView: GrabSongWishView? = null
    private var mEditRoomDialog: DialogPlus? = null

    lateinit var mOwnerManagePresenter: GrabSongManagePresenter
    private var mRoomData: GrabRoomData? = null

    var mTagModelList: List<RecommendTagModel>? = null

    override fun initView(): Int {
        return R.layout.grab_song_manage_fragment_layout
    }

    override fun initData(savedInstanceState: Bundle?) {
        if (mRoomData == null) {
            if (activity != null) {
                activity!!.finish()
            }
            return
        }

        mCommonTitleBar = mRootView.findViewById<View>(R.id.titlebar) as CommonTitleBar
        mSearchSongIv = mRootView.findViewById<View>(R.id.search_song_iv) as ExTextView
        mTagTab = mRootView.findViewById<View>(R.id.tag_tab) as SlidingTabLayout
        mViewpager = mRootView.findViewById<View>(R.id.viewpager) as ViewPager

        mCommonTitleBar.centerTextView.setOnClickListener(object : DebounceViewClickListener() {
            override fun clickValid(v: View) {
                showEditRoomDialog()
            }
        })

        mCommonTitleBar.leftTextView.setOnClickListener(object : DebounceViewClickListener() {
            override fun clickValid(v: View) {
                if (activity is SongManagerActivity) {
                    if (activity != null) {
                        activity!!.finish()
                    }
                } else {
                    finish()
                }
            }
        })

        mOwnerManagePresenter = GrabSongManagePresenter(this, mRoomData)
        addPresent(mOwnerManagePresenter)
        mOwnerManagePresenter.getRecommendTag()
        showRoomName(mRoomData!!.roomName)

        mSearchSongIv.setOnClickListener(object : DebounceViewClickListener() {
            override fun clickValid(v: View) {
                U.getFragmentUtils().addFragment(FragmentUtils.newAddParamsBuilder(context as BaseActivity?, GrabSearchSongFragment::class.java)
                        .setAddToBackStack(true)
                        .setHasAnimation(true)
                        .addDataBeforeAdd(0, SongManagerActivity.TYPE_FROM_GRAB)
                        .addDataBeforeAdd(1, mRoomData!!.isOwner)
                        .setFragmentDataListener { requestCode, resultCode, bundle, obj ->
                            if (requestCode == 0 && resultCode == 0 && obj != null) {
                                val model = obj as SongModel
                                MyLog.d(TAG, "onFragmentResult model=$model")
                                EventBus.getDefault().post(AddSongEvent(model))
                            }
                        }
                        .build())
            }
        })
    }

    override fun showRoomName(roomName: String) {
        mCommonTitleBar.centerTextView.text = roomName
        if (mRoomData!!.isOwner) {
            val drawable = U.getDrawable(R.drawable.ycdd_edit_roomname_icon)
            drawable.bounds = Rect(0, 0, drawable.intrinsicWidth, drawable.intrinsicHeight)
            mCommonTitleBar.centerTextView.setCompoundDrawables(null, null, drawable, null)
            mCommonTitleBar.centerTextView.compoundDrawablePadding = U.getDisplayUtils().dip2px(7f)
            mCommonTitleBar.centerTextView.isClickable = true
        } else {
            mCommonTitleBar.centerTextView.isClickable = false
        }
    }

    /**
     * 得到所有类别
     *
     * @param recommendTagModelList
     */

    override fun showRecommendSong(recommendTagModelList: MutableList<RecommendTagModel>) {
        if (recommendTagModelList == null || recommendTagModelList.size == 0) {
            return
        }


        if (mRoomData!!.isOwner) {
            val recommendModel = RecommendTagModel()
            recommendModel.type = -1
            recommendModel.name = "愿望歌单"
            recommendTagModelList.add(0, recommendModel)

            val recommendTagModel = RecommendTagModel()
            recommendTagModel.type = -1
            recommendTagModel.name = "已点0"
            recommendTagModelList.add(0, recommendTagModel)
        }

        mTagModelList = recommendTagModelList
        mTagTab.setCustomTabView(R.layout.manage_song_tab, R.id.tab_tv)
        mTagTab.setSelectedIndicatorColors(U.getColor(R.color.black_trans_20))
        mTagTab.setDistributeMode(SlidingTabLayout.DISTRIBUTE_MODE_NONE)
        mTagTab.setIndicatorAnimationMode(SlidingTabLayout.ANI_MODE_NORMAL)
        mTagTab.setSelectedIndicatorThickness(U.getDisplayUtils().dip2px(24f).toFloat())
        mTagTab.setIndicatorCornorRadius(U.getDisplayUtils().dip2px(12f).toFloat())
        mPagerAdapter = object : PagerAdapter() {

            override fun destroyItem(container: ViewGroup, position: Int, `object`: Any) {
                container.removeView(`object` as View)
            }

            override fun instantiateItem(container: ViewGroup, position: Int): Any {
                return instantiateItemGrab(container, position, mTagModelList!!)
            }

            override fun getCount(): Int {
                return mTagModelList!!.size
            }

            override fun isViewFromObject(view: View, `object`: Any): Boolean {
                return view === `object`
            }

            override fun getPageTitle(position: Int): CharSequence? {
                return mTagModelList!![position].name
            }
        }

        mTagTab.setOnPageChangeListener(object : ViewPager.OnPageChangeListener {
            override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {

            }

            override fun onPageSelected(position: Int) {
                val view = mViewpager.findViewWithTag<View>(position)
                if (view != null) {
                    if (view is RecommendSongView) {
                        view.tryLoad()
                    } else if (view is GrabSongWishView) {
                        view.tryLoad()
                    } else if (view is GrabExistSongManageView) {
                        view.tryLoad()
                    }
                }
            }

            override fun onPageScrollStateChanged(state: Int) {

            }
        })

        mViewpager.adapter = mPagerAdapter
        mTagTab.setViewPager(mViewpager)
        mPagerAdapter.notifyDataSetChanged()
    }

    fun instantiateItemGrab(container: ViewGroup, position: Int, recommendTagModelList: List<RecommendTagModel>): Any {
        MyLog.d(TAG, "instantiateItem container=$container position=$position")
        var view: View

        if (mRoomData!!.isOwner) {
            if (position == 0) {
                if (mGrabSongManageView == null) {
                    mGrabSongManageView = GrabExistSongManageView(context!!, mRoomData!!)
                }
                view = mGrabSongManageView!!
            } else if (position == 1) {
                if (mGrabSongWishView == null) {
                    mGrabSongWishView = GrabSongWishView(context!!, mRoomData!!)
                }
                mGrabSongWishView!!.tag = position
                view = mGrabSongWishView!!
            } else {
                val recommendTagModel = recommendTagModelList[position]
                val recommendSongView = RecommendSongView(activity!!, SongManagerActivity.TYPE_FROM_GRAB,
                        mRoomData!!.isOwner, mRoomData!!.gameId, recommendTagModel)
                recommendSongView.tag = position
                view = recommendSongView
            }
        } else {
            val recommendTagModel = recommendTagModelList[position]
            val recommendSongView = RecommendSongView(activity!!, SongManagerActivity.TYPE_FROM_GRAB,
                    mRoomData!!.isOwner, mRoomData!!.gameId, recommendTagModel)
            recommendSongView.tag = position
            view = recommendSongView
        }

        if (container.indexOfChild(view) == -1) {
            container.addView(view)
        }

        return view
    }

    override fun setData(type: Int, data: Any?) {
        super.setData(type, data)
        if (type == 0) {
            mRoomData = data as GrabRoomData?
        }
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onEvent(event: SongNumChangeEvent) {
        mTagModelList?.get(0)?.name = "已点" + event.songNum
        mTagTab.notifyDataChange()
    }

    private fun showEditRoomDialog() {
        val grabEditView = GrabEditRoomNameView(context!!, mRoomData!!.roomName)
        grabEditView.onClickCancel = {
            mEditRoomDialog?.dismiss()
        }
        grabEditView.onClickSave = {
            if (!TextUtils.isEmpty(it)) {
                // TODO: 2019/4/18 修改房间名
                mEditRoomDialog?.dismiss(false)
                mOwnerManagePresenter.updateRoomName(mRoomData!!.gameId, it)
            } else {
                // TODO: 2019/4/18 房间名为空
                U.getToastUtil().showShort("输入的房间名为空")
            }
        }

        mEditRoomDialog?.dismiss(false)
        mEditRoomDialog = DialogPlus.newDialog(context!!)
                .setContentHolder(ViewHolder(grabEditView))
                .setContentBackgroundResource(R.color.transparent)
                .setOverlayBackgroundResource(R.color.black_trans_50)
                .setExpanded(false)
                .setGravity(Gravity.BOTTOM)
                .setOnDismissListener { U.getKeyBoardUtils().hideSoftInputKeyBoard(activity) }
                .create()
        U.getKeyBoardUtils().showSoftInputKeyBoard(activity)
        mEditRoomDialog!!.show()
    }

    override fun destroy() {
        super.destroy()
        //        for (RecommendSongView recommendSongView : mRecommendSongViews) {
        //            recommendSongView.destroy();
        //        }

        if (mGrabSongManageView != null) {
            mGrabSongManageView!!.destroy()
        }

        if (mGrabSongWishView != null) {
            mGrabSongWishView!!.destroy()
        }
    }

    override fun onBackPressed(): Boolean {
        if (mEditRoomDialog != null && mEditRoomDialog!!.isShowing) {
            mEditRoomDialog!!.dismiss(false)
            mEditRoomDialog = null
            return true
        }
        if (activity != null) {
            activity!!.finish()
        }
        return true
    }

    override fun useEventBus(): Boolean {
        return true
    }
}
